backend/src/indexer/block_preprocessor.rs

```diff
@@ -281,49 +300,106 @@ impl concordium_rust_sdk::indexer::Indexer for BlockPreProcessor {
     }
 }
 
-/// compute validator staking information such as the validators stake and the
-/// total staked in their pool by getting pool info. pool info did not exist
-/// prior to protocol version 4, in this case the account info is used to get
-/// the stake for the validator and the total pool for the validator prior to
-/// protocol version 4 will be set to their account stake.
+/// Responsible for computing the latest validator staking information which
+/// returns a Tuple of <Amount, ValidatorStakingInformation> - where Amount is
+/// the total stake summed for all the validators
+/// and the ValidatorStakingInformation is a struct containing all of the new
+/// information we need to update for the validators such as their staked
+/// amounts and their pool staked amounts. As it is intensive to go to the node
+/// through the client, the last computed total stake is stored in a cache. We
+/// only go to the retrieve the latest stakes if enough blocks have passed
+/// (defined by our env variable STAKE_RECOMPUTE_INTERVAL_IN_BLOCKS). It is
+/// important to note that in our cached scenario, we just return the last
+/// cached total stake amount - we do not return validatorStakingInformation
+/// because: if this is present when we reach the block processor it will
+/// perform database update queries for all the validator information provided,
+/// therefore it is crucial to return a struct that represents 'no updates at
+/// this time'
```

Comment:
The `compute_validator_staking_information` is primarily executed in the `parallel` stage of the indexer. That is properly why you needed to add the `readWrite` lock because several processes tried to update the value. It seems a wrong design to now force the `parallel` stage to do sequential updates with a `ReadWrite` cache as we loose the parallel behavior. I wonder if we should move the `compute_validator_staking_information` to the save stage so that we can avoid the  `readWrite` lock.